### ISSUE TITLE
Laws.hs lacking some Additive contexts

### ISSUE THREAD

**Issue by Magalame** [tonyday567/numhask#133]
Hi again!
I tried adding `numhask-prelude` to a project with lts-12.14, and I ran into this compiling error:

```
--  While building package numhask-prelude-0.0.5.1 using:
      /home/user/.stack/setup-exe-cache/x86_64-linux/Cabal-simple_mPHDZzAJ_2.2.0.1_ghc-8.4.3 --builddir=.stack-work/dist/x86_64-linux/Cabal-2.2.0.1 build --ghc-options " -ddump-hi -ddump-to-file -fdiagnostics-color=always"
    Process exited with code: ExitFailure 1
    Logs have been written to: /home/user/Projects/fastest-matrices/.stack-work/logs/numhask-prelude-0.0.5.1.log

    Configuring numhask-prelude-0.0.5.1...
    Preprocessing library for numhask-prelude-0.0.5.1..
    Building library for numhask-prelude-0.0.5.1..
    [1 of 5] Compiling NumHask.Error    ( src/NumHask/Error.hs, .stack-work/dist/x86_64-linux/Cabal-2.2.0.1/build/NumHask/Error.o )
    [2 of 5] Compiling NumHask.Prelude  ( src/NumHask/Prelude.hs, .stack-work/dist/x86_64-linux/Cabal-2.2.0.1/build/NumHask/Prelude.o )
    [3 of 5] Compiling NumHask.Laws     ( src/NumHask/Laws.hs, .stack-work/dist/x86_64-linux/Cabal-2.2.0.1/build/NumHask/Laws.o )
    
    /tmp/stack9181/numhask-prelude-0.0.5.1/src/NumHask/Laws.hs:142:26: error:
        • Could not deduce (Additive a) arising from a use of ‘+’
          from the context: Epsilon a
            bound by the type signature for:
                       additiveLaws_ :: forall a. Epsilon a => [Law a]
            at src/NumHask/Laws.hs:139:1-39
          Possible fix:
            add (Additive a) to the context of
              the type signature for:
                additiveLaws_ :: forall a. Epsilon a => [Law a]
        • In the first argument of ‘(≈)’, namely ‘(a + b) + c’
          In the expression: (a + b) + c ≈ a + (b + c)
          In the first argument of ‘Ternary’, namely
            ‘(\ a b c -> (a + b) + c ≈ a + (b + c))’
        |
    142 |     , Ternary (\a b c -> (a + b) + c ≈ a + (b + c)))
        |                          ^^^^^^^^^^^
    
    /tmp/stack9181/numhask-prelude-0.0.5.1/src/NumHask/Laws.hs:301:18: error:
        • Could not deduce (Additive b) arising from a use of ‘+’
          from the context: (FromInteger b, Ord b, Signed b, Epsilon b,
                             Metric a b)
            bound by the type signature for:
                       metricIntegralLaws :: forall a b.
                                             (FromInteger b, Ord b, Signed b, Epsilon b, Metric a b) =>
                                             [Law2 a b]
            at src/NumHask/Laws.hs:(283,1)-(284,12)
          Possible fix:
            add (Additive b) to the context of
              the type signature for:
                metricIntegralLaws :: forall a b.
                                      (FromInteger b, Ord b, Signed b, Epsilon b, Metric a b) =>
                                      [Law2 a b]
        • In the first argument of ‘(-)’, namely
            ‘distanceLp p a c + distanceLp p b c’
          In the first argument of ‘veryNegative’, namely
            ‘(distanceLp p a c + distanceLp p b c - distanceLp p a b)’
          In the first argument of ‘not’, namely
            ‘(veryNegative
                (distanceLp p a c + distanceLp p b c - distanceLp p a b))’
        |
    301 |                 (distanceLp p a c + distanceLp p b c - distanceLp p a b)) &&
        |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    
    /tmp/stack9181/numhask-prelude-0.0.5.1/src/NumHask/Laws.hs:301:18: error:
        • Could not deduce (AdditiveGroup b) arising from a use of ‘-’
          from the context: (FromInteger b, Ord b, Signed b, Epsilon b,
                             Metric a b)
            bound by the type signature for:
                       metricIntegralLaws :: forall a b.
                                             (FromInteger b, Ord b, Signed b, Epsilon b, Metric a b) =>
                                             [Law2 a b]
            at src/NumHask/Laws.hs:(283,1)-(284,12)
          Possible fix:
            add (AdditiveGroup b) to the context of
              the type signature for:
                metricIntegralLaws :: forall a b.
                                      (FromInteger b, Ord b, Signed b, Epsilon b, Metric a b) =>
                                      [Law2 a b]
        • In the first argument of ‘veryNegative’, namely
            ‘(distanceLp p a c + distanceLp p b c - distanceLp p a b)’
          In the first argument of ‘not’, namely
            ‘(veryNegative
                (distanceLp p a c + distanceLp p b c - distanceLp p a b))’
          In the first argument of ‘(&&)’, namely
            ‘not
               (veryNegative
                  (distanceLp p a c + distanceLp p b c - distanceLp p a b))’
        |
    301 |                 (distanceLp p a c + distanceLp p b c - distanceLp p a b)) &&
        |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    
    /tmp/stack9181/numhask-prelude-0.0.5.1/src/NumHask/Laws.hs:347:18: error:
        • Could not deduce (Additive b) arising from a use of ‘+’
          from the context: (FromRatio b, Ord b, Signed b, Epsilon b,
                             Metric a b, Normed a b)
            bound by the type signature for:
                       metricRationalLaws :: forall a b.
                                             (FromRatio b, Ord b, Signed b, Epsilon b, Metric a b,
                                              Normed a b) =>
                                             [Law2 a b]
            at src/NumHask/Laws.hs:(326,1)-(327,12)
          Possible fix:
            add (Additive b) to the context of
              the type signature for:
                metricRationalLaws :: forall a b.
                                      (FromRatio b, Ord b, Signed b, Epsilon b, Metric a b,
                                       Normed a b) =>
                                      [Law2 a b]
        • In the first argument of ‘(-)’, namely
            ‘distanceLp p a c + distanceLp p b c’
          In the first argument of ‘veryNegative’, namely
            ‘(distanceLp p a c + distanceLp p b c - distanceLp p a b)’
          In the first argument of ‘not’, namely
            ‘(veryNegative
                (distanceLp p a c + distanceLp p b c - distanceLp p a b))’
        |
    347 |                 (distanceLp p a c + distanceLp p b c - distanceLp p a b)) &&
        |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    
    /tmp/stack9181/numhask-prelude-0.0.5.1/src/NumHask/Laws.hs:347:18: error:
        • Could not deduce (AdditiveGroup b) arising from a use of ‘-’
          from the context: (FromRatio b, Ord b, Signed b, Epsilon b,
                             Metric a b, Normed a b)
            bound by the type signature for:
                       metricRationalLaws :: forall a b.
                                             (FromRatio b, Ord b, Signed b, Epsilon b, Metric a b,
                                              Normed a b) =>
                                             [Law2 a b]
            at src/NumHask/Laws.hs:(326,1)-(327,12)
          Possible fix:
            add (AdditiveGroup b) to the context of
              the type signature for:
                metricRationalLaws :: forall a b.
                                      (FromRatio b, Ord b, Signed b, Epsilon b, Metric a b,
                                       Normed a b) =>
                                      [Law2 a b]
        • In the first argument of ‘veryNegative’, namely
            ‘(distanceLp p a c + distanceLp p b c - distanceLp p a b)’
          In the first argument of ‘not’, namely
            ‘(veryNegative
                (distanceLp p a c + distanceLp p b c - distanceLp p a b))’
          In the first argument of ‘(&&)’, namely
            ‘not
               (veryNegative
                  (distanceLp p a c + distanceLp p b c - distanceLp p a b))’
        |
    347 |                 (distanceLp p a c + distanceLp p b c - distanceLp p a b)) &&
        |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    
    /tmp/stack9181/numhask-prelude-0.0.5.1/src/NumHask/Laws.hs:382:13: error:
        • Could not deduce (Ord a) arising from a use of ‘<’
          from the context: (Field a, QuotientField a Integer, FromInteger a)
            bound by the type signature for:
                       quotientFieldLaws :: forall a.
                                            (Field a, QuotientField a Integer, FromInteger a) =>
                                            [Law2 a Integer]
            at src/NumHask/Laws.hs:377:1-90
          Possible fix:
            add (Ord a) to the context of
              the type signature for:
                quotientFieldLaws :: forall a.
                                     (Field a, QuotientField a Integer, FromInteger a) =>
                                     [Law2 a Integer]
        • In the first argument of ‘(&&)’, namely
            ‘((a - one) < (fromInteger (floor a)))’
          In the expression:
            ((a - one) < (fromInteger (floor a)))
              &&
                (fromInteger (floor a) <= a)
                  &&
                    (a <= fromInteger (ceiling a))
                      && (fromInteger (ceiling a) < a + one)
          In the first argument of ‘Unary10’, namely
            ‘(\ a
                -> ((a - one) < (fromInteger (floor a)))
                     &&
                       (fromInteger (floor a) <= a)
                         &&
                           (a <= fromInteger (ceiling a))
                             && (fromInteger (ceiling a) < a + one))’
        |
    382 |            ((a - one) < (fromInteger (floor a))) &&
        |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    
    /tmp/stack9181/numhask-prelude-0.0.5.1/src/NumHask/Laws.hs:453:29: error:
        • Could not deduce (Additive (r a)) arising from a use of ‘+’
          from the context: (Epsilon a, Epsilon (r a), AdditiveModule r a)
            bound by the type signature for:
                       additiveModuleLaws :: forall a (r :: * -> *).
                                             (Epsilon a, Epsilon (r a), AdditiveModule r a) =>
                                             [Law2 (r a) a]
            at src/NumHask/Laws.hs:(449,1)-(450,69)
        • In the first argument of ‘(.+)’, namely ‘(a + b)’
          In the first argument of ‘(≈)’, namely ‘(a + b) .+ c’
          In the expression: (a + b) .+ c ≈ a + (b .+ c)
        |
    453 |     , Ternary21 (\a b c -> (a + b) .+ c ≈ a + (b .+ c)))
        |                             ^^^^^
    
    /tmp/stack9181/numhask-prelude-0.0.5.1/src/NumHask/Laws.hs:466:29: error:
        • Could not deduce (Additive (r a)) arising from a use of ‘+’
          from the context: (Epsilon a, Epsilon (r a),
                             AdditiveGroupModule r a)
            bound by the type signature for:
                       additiveGroupModuleLaws :: forall a (r :: * -> *).
                                                  (Epsilon a, Epsilon (r a), AdditiveGroupModule r a) =>
                                                  [Law2 (r a) a]
            at src/NumHask/Laws.hs:(461,1)-(463,19)
        • In the first argument of ‘(.-)’, namely ‘(a + b)’
          In the first argument of ‘(≈)’, namely ‘(a + b) .- c’
          In the expression: (a + b) .- c ≈ a + (b .- c)
        |
    466 |     , Ternary21 (\a b c -> (a + b) .- c ≈ a + (b .- c)))
        |                             ^^^^^
    
    /tmp/stack9181/numhask-prelude-0.0.5.1/src/NumHask/Laws.hs:482:29: error:
        • Could not deduce (Additive (r a)) arising from a use of ‘+’
          from the context: (Epsilon a, Epsilon (r a),
                             MultiplicativeModule r a)
            bound by the type signature for:
                       multiplicativeModuleLaws :: forall a (r :: * -> *).
                                                   (Epsilon a, Epsilon (r a),
                                                    MultiplicativeModule r a) =>
                                                   [Law2 (r a) a]
            at src/NumHask/Laws.hs:(475,1)-(477,19)
        • In the first argument of ‘(.*)’, namely ‘(a + b)’
          In the first argument of ‘(≈)’, namely ‘(a + b) .* c’
          In the expression: (a + b) .* c ≈ (a .* c) + (b .* c)
        |
    482 |     , Ternary21 (\a b c -> (a + b) .* c ≈ (a .* c) + (b .* c)))
        |                             ^^^^^
    
    /tmp/stack9181/numhask-prelude-0.0.5.1/src/NumHask/Laws.hs:542:35: error:
        • Could not deduce (Additive (r a)) arising from a use of ‘+’
          from the context: (MultiplicativeModule r a, Epsilon a,
                             Epsilon (r a), Hilbert r a)
            bound by the type signature for:
                       hilbertLaws :: forall (r :: * -> *) a.
                                      (MultiplicativeModule r a, Epsilon a, Epsilon (r a),
                                       Hilbert r a) =>
                                      [Law2 (r a) a]
            at src/NumHask/Laws.hs:(533,1)-(538,19)
        • In the second argument of ‘(<.>)’, namely ‘(b + c)’
          In the first argument of ‘(≈)’, namely ‘a <.> (b + c)’
          In the expression: a <.> (b + c) ≈ a <.> b + a <.> c
        |
    542 |     , Ternary30 (\a b c -> a <.> (b + c) ≈ a <.> b + a <.> c))
        |                                   ^^^^^
    
    /tmp/stack9181/numhask-prelude-0.0.5.1/src/NumHask/Laws.hs:558:48: error:
        • Could not deduce (Additive (r a)) arising from a use of ‘+’
          from the context: (Eq (r (r a)), Additive (r (r a)),
                             TensorProduct (r a), Epsilon (r a))
            bound by the type signature for:
                       tensorProductLaws :: forall (r :: * -> *) a.
                                            (Eq (r (r a)), Additive (r (r a)), TensorProduct (r a),
                                             Epsilon (r a)) =>
                                            [Law2 (r a) a]
            at src/NumHask/Laws.hs:(549,1)-(555,19)
        • In the first argument of ‘(><)’, namely ‘(a + c)’
          In the second argument of ‘(==)’, namely ‘(a + c) >< b’
          In the expression: a >< b + c >< b == (a + c) >< b
        |
    558 |     , Ternary30 (\a b c -> a >< b + c >< b == (a + c) >< b))
        |                                                ^^^^^
    
```

**Comment by Magalame** [comment]
solved by using `numhask-array-0.3`

**Comment by tonyday567** [comment]
I just finished uploading the latest numhask libraries to hackage, so the starting point for more stable support is:

- numhask-0.3.0.0
- numhask-array-0.3
- numhask-prelude-0.3.1
- numhask-space-0.1.1
- numhask-hedgehog-0.3

I've also added numhask to the stackage nightlies.  Will wait a bit and see it's ok, then slowly add the others over the next few days. It should all be compatible back to lts-12.14, but will start to get wonky before that. 

Thanks for the report!
